### PR TITLE
Milestone 2 status

### PR DESCRIPTION
**Pull request by paweld2**
The servlet integration is already started and the login and logout actions send requests to castle.

### CODE REVIEW

File: src/main/java/io/castle/example/LoginServlet.java [hunk 1]
```diff
@@ -1,38 +1,83 @@
...
+
     @Override
     protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
-        Castle.setAPISecret(Config.apiKey());
+        CastleApi castleApi = Castle.sdk().onRequest(req);
 
-        Event event = new Event();
-        event.setUserId("1234");
-        event.setName("$login.succeeded");
+        String overrideString = req.getParameter("override");
```
Comment (lluft) [inline src/main/java/io/castle/example/LoginServlet.java]: I'm guessing you added this for test purposes. please remove the override functionality

File: src/main/java/io/castle/example/LoginServlet.java [hunk 2]
```diff
@@ -1,38 +1,83 @@
...
+                }
+                break;
+            }
+        } else
+
+        {
+            if (user != null) {
+                castleApi.track("$login.failed", user.getId().toString());
+            } else {
+                castleApi.track("$login.failed");
```
Comment (lluft) [inline src/main/java/io/castle/example/LoginServlet.java]: add the properties to login failed. see ruby example
```ruby
castle.track(
  name: '$login.failed',
  user_id: '1234', # Omit this if no matching user was found
  properties: {
    email: 'johan@castle.io'
  })
```

Comment (lluft) [reply src/main/java/io/castle/example/LoginServlet.java]: in case of `user == null`

File: src/main/java/io/castle/example/LogoutServlet.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        CastleApi castleApi = Castle.sdk().onRequest(req);
+        HttpSession session = req.getSession(true);
+        Object userObject = session.getAttribute("currentSessionUser");
+        TestUser user = (TestUser) userObject;
+
+        //Notice that this is a custom event.
+        CustomProperties customExample = new CustomProperties();
+        customExample.setExampleValue("valueToSend");
+        castleApi.track("logout", user.getId().toString(), customExample);
```
Comment (lluft) [inline src/main/java/io/castle/example/LogoutServlet.java]: should be `$logout.succeeded`

File: src/main/java/io/castle/example/PasswordResetRequiredServlet.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+        CastleApi castleApi = Castle.sdk().onRequest(req);
+        String login = req.getParameter("login");
+        TestUser user = UserAuthenticationBackend.findUser(login);
+        if (user != null) {
+            castleApi.track("$password_reset_request.succeeded", user.getId().toString());
+            session.setAttribute("passwordResetUser", user);
+            resp.sendRedirect("password_reset_request_succeeded.jsp");
+        } else {
+            castleApi.track("$password_reset_request.failed");
```
Comment (lluft) [inline src/main/java/io/castle/example/PasswordResetRequiredServlet.java]: as with `$login.failed` send `login` in the properties in track. depending on what you expect login to be send `username` or `email` as key in the properties

File: src/main/webapp/email_change_form.jsp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    <c:when test="${currentUser != null }">
+        <html>
+        <head>
+            <script type="text/javascript">
+                <jsp:include page="_castle_script.jsp"/>
+//                secure mode off
+                _castle('setUser', {
+                    id: '<c:out value = "${currentUser}"/>'
+                });
+//                secure mod on
```
Comment (lluft) [inline src/main/webapp/email_change_form.jsp]: mode

File: src/main/java/io/castle/example/SetupListener.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+public class SetupListener implements ServletContextListener {
+
+    /**
+     * Example of how to verify the SDK configuration during application initialization to avoid errors during request handling.
+     *
+     * @param sce
+     */
+    public void contextInitialized(ServletContextEvent sce) {
+        try {
+            Castle.verifySdkConfigurationAndInitialize();
```
Comment (lluft) [inline src/main/java/io/castle/example/SetupListener.java:19]: could you extend the example to changing something in the configuration. e.g. add a header to the blacklist

**Review by lluft — CHANGES_REQUESTED** [review]

File: src/main/java/io/castle/example/SetupListener.java (hunk 6 above)
Comment (paweld2) [reply src/main/java/io/castle/example/SetupListener.java:19]: The configuration file contains additional headers for the blacklist, please check the file:
src/main/resources/castle_sdk.properties

File: src/main/java/io/castle/example/LoginServlet.java (hunk 1 above)
Comment (paweld2) [reply src/main/java/io/castle/example/LoginServlet.java]: Done.

File: src/main/java/io/castle/example/LoginServlet.java (hunk 2 above)
Comment (paweld2) [reply src/main/java/io/castle/example/LoginServlet.java]: done.

File: src/main/java/io/castle/example/LogoutServlet.java (hunk 3 above)
Comment (paweld2) [reply src/main/java/io/castle/example/LogoutServlet.java]: done.

File: src/main/java/io/castle/example/PasswordResetRequiredServlet.java (hunk 4 above)
Comment (paweld2) [reply src/main/java/io/castle/example/PasswordResetRequiredServlet.java]: done.

File: src/main/webapp/email_change_form.jsp (hunk 5 above)
Comment (paweld2) [reply src/main/webapp/email_change_form.jsp]: done

**Review by lluft — APPROVED** [review]

**Pull request merged**